scipy/stats/_survival.py

```diff
@@ -374,3 +391,104 @@ def _ecdf_right_censored(sample):
     sf = np.cumprod((n - d) / n)
     cdf = 1 - sf
     return t, cdf, sf, n, d
+
+
+@dataclass
+class LogRankResult:
+    statistic: np.ndarray
+    pvalue: np.ndarray
+
+
+def log_rank(
+    x: npt.ArrayLike | CensoredData,
+    y: npt.ArrayLike | CensoredData
+) -> LogRankResult:
+    """Compare the survival distributions of two samples via the logrank test.
+
+    Parameters
+    ----------
+    x, y : array_like or CensoredData
+        Samples to compare based on their survival functions.
+
+    Returns
+    -------
+    res : LogRankResult
+        An object containing attributes:
+
+        statistic : float ndarray
+            The computed statistic of the test.
+        pvalue : float ndarray
+            The computed p-value of the test.
+
+    References
+    ----------
+    .. [1] Peto, Richard and Peto, Julian. "Asymptotically Efficient Rank
+           Invariant Test Procedures." J. R. Statist. Soc. A, 185 (1972)
+
+    Examples
+    --------
+
+    >>> from scipy import stats
+    >>> x = stats.CensoredData(
+    ...     uncensored=[8, 12, 26, 14, 21, 27],
+    ...     right=[8, 32, 20, 40]
+    ... )
+    >>> y = stats.CensoredData(
+    ...     uncensored=[33, 28, 41],
+    ...     right=[48, 48, 25, 37, 48, 25, 43]
+    ... )
+
+    >>> import numpy as np
+    >>> import matplotlib.pyplot as plt
+    >>> ax = plt.subplot()
+    >>> ecdf_sample = stats.ecdf(x)
+    >>> ax.step(np.insert(ecdf_sample.x, 0, 0), np.insert(ecdf_sample.sf.points, 0, 1),
+    ...         where='post', label='sample')
+    >>> ecdf_control = stats.ecdf(y)
+    >>> ax.step(np.insert(ecdf_control.x, 0, 0), np.insert(ecdf_control.sf.points, 0, 1),
+    ...         ls='-.', where='post', label='control')
+    >>> ax.set_xlabel('Time (months)')
+    >>> ax.set_ylabel('Empirical CDF')
+    >>> plt.legend()
+    >>> plt.show()
+
+    >>> res = stats.log_rank(x=x, y=y)
+    >>> res.statistic
+    6.148087536256203
+    >>> res.pvalue
+    0.013155428547469983
+
+    """
+    x = _iv_CensoredData(sample=x, param_name='x')
+    y = _iv_CensoredData(sample=y, param_name='y')
+
+    xy = CensoredData(
+        uncensored=np.concatenate((x._uncensored, y._uncensored)),
+        right=np.concatenate((x._right, y._right))
+    )
+
+    res = ecdf(xy)
+    idx = res.sf._d.astype(bool)  # events were observed (not censored)
+    times_xy = res.x[idx]
+    at_risk_xy = res.sf._n[idx]
+    deaths_xy = res.sf._d[idx]
+
+    n_died_x = x._uncensored.size
+    n_died_y = y._uncensored.size
+
+    res_x = ecdf(x)
+    i = np.searchsorted(res_x.x, times_xy)  # or use `interpolate_1d`
+    at_risk_x = np.append(res_x.sf._n, 0)[i]  # 0 at risk after last x time
+    at_risk_y = at_risk_xy - at_risk_x
+
+    sum_exp_event_sample = np.sum(at_risk_x * (deaths_xy/at_risk_xy))
+    sum_exp_event_control = np.sum(at_risk_y * (deaths_xy/at_risk_xy))
```

Comment:
Consistency "event"/"deaths", "sample"/"x", etc.
```suggestion
    sum_exp_deaths_x = np.sum(at_risk_x * (deaths_xy/at_risk_xy))
    sum_exp_deaths_y = np.sum(at_risk_y * (deaths_xy/at_risk_xy))
```